File: packages/sdk-python/tests/integration_test.py
# -*- coding: utf-8 -*-
"""Python SDK 联调：对着真实服务端跑完整流程。"""
import os, shutil, sys, json, time
sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
from jc_kami import LicenseClient, LicenseError, ErrorCode

env = {}
for line in open("/tmp/sdk.env"):
    k, _, v = line.strip().partition("=")
    env[k] = v

RUN = os.urandom(4).hex()          # 每次运行用独立的设备与目录，保证可重入
STORE = "/tmp/py-sdk-state-" + RUN
shutil.rmtree(STORE, ignore_errors=True)

policies = []
def make(version="1.0.0", store=STORE, **kw):
    return LicenseClient(
        app_id="sdk_demo", server_url="http://127.0.0.1:3100",
        plugin_id=env["SDK_PLUGIN"], plugin_token=env["SDK_TOKEN"],
        plugin_secret=env["SDK_SECRET"], client_version=version,
        storage_dir=store, on_policy=lambda p: policies.append(p), **kw)

ok = lambda m: print("  \033[32m✓\033[0m " + m)
bad = lambda m: (print("  \033[31m✗ " + m + "\033[0m"), sys.exit(1))

print("\n【1】未激活时 verify")
c = make()
print("     设备指纹:", c.device_id[:16] + "...")
try:
    c.verify(); bad("应当抛错")
except LicenseError as e:
    ok("%s — %s" % (e.code, e.message)) if e.code == ErrorCode.NOT_ACTIVATED_LOCALLY else bad(e.code)

print("\n【2】拉取策略（未激活也能调）")
p = c.fetch_policy()
ok("app_status=%s  最新版本=%s" % (p["app_status"], p["upgrade"]["latest_version"]))

print("\n【3】激活")
r = c.activate(env["SDK_FRESH1"])
ok("license_id=%s  到期=%s  设备 %d/%d  离线宽限 %dh" % (
    r["license_id"], r["expires_at"][:10], r["device_count"], r["device_limit"], r["offline_grace_hours"]))

print("\n【4】启动校验（令牌滚动续期）")
t1 = r["license_token"]
r2 = c.verify()
ok("验证通过，换发了新令牌" if r2["license_token"] != t1 else "令牌未变化（不符合预期）")

print("\n【5】本地已存凭据")
ok("has_local_license() = %s" % c.has_local_license())

print("\n【6】错误卡密")
try:
    make(store="/tmp/py-sdk-s2-" + RUN).activate("ZZZZZ-ZZZZZ-ZZZZZ-ZZZZZ-ZZZZZ"); bad("应当抛错")
except LicenseError as e:
    ok("%s — %s" % (e.code, e.message)) if e.code == ErrorCode.LICENSE_NOT_FOUND else bad(e.code)

print("\n【7】超出设备上限（上限 2）")

c2 = make(store="/tmp/py-sdk-d2-" + RUN, device_id="py-dev2-" + RUN)
c2.activate(env["SDK_FRESH1"]); ok("第 2 台设备绑定成功")
c3 = make(store="/tmp/py-sdk-d3-" + RUN, device_id="py-dev3-" + RUN)
try:
    c3.activate(env["SDK_FRESH1"]); bad("应当被拒")
except LicenseError as e:
    ok("%s — %s" % (e.code, e.message)) if e.code == ErrorCode.DEVICE_LIMIT_EXCEEDED else bad(e.code)

print("\n【8】主动解绑（消耗换绑配额）")
d = c2.deactivate("换机器")
ok("已解绑，剩余换绑 %s" % d.get("rebind_remaining"))

print("\n【9】本地文件防篡改")
path = os.path.join(STORE, "license.dat")
raw = bytearray(open(path, "rb").read())
raw[-20] ^= 0xFF          # 改动 payload 里的一个字节
open(path, "wb").write(bytes(raw))
try:
    c.verify(); bad("篡改未被发现")
except LicenseError as e:
    ok("%s — %s" % (e.code, e.message)) if e.code == ErrorCode.LOCAL_STORAGE_TAMPERED else bad(e.code)

print("\n【10】离线宽限期")
shutil.rmtree(STORE, ignore_errors=True)
c = make(); c.activate(env["SDK_FRESH2"])
c._storage.write({**c._storage.read(), "last_verified_at": "2020-01-01T00:00:00Z"})
c.server_url = "http://127.0.0.1:59999"   # 指向一个连不上的端口，模拟断网
try:
    c.verify(); bad("超期仍放行")
except LicenseError as e:
    ok("超过宽限期被拒: %s — %s" % (e.code, e.message)) if e.code == ErrorCode.NETWORK_UNAVAILABLE else bad(e.code)

st = c._storage.read(); st["last_verified_at"] = __import__("datetime").datetime.now(
    __import__("datetime").timezone.utc).isoformat().replace("+00:00", "Z")
c._storage.write(st)
r = c.verify()
ok("宽限期内离线放行: offline=%s" % r["offline"]) if r.get("offline") else bad("未走离线分支")

print("\n【11】状态查询（不签发令牌）")
c.server_url = "http://127.0.0.1:3100"
s = c.status()
ok("status=%s  本机已绑定=%s  设备 %d/%d" % (s["status"], s["device_bound"], s["device_count"], s["device_limit"]))
if "license_token" in s: bad("状态查询不应返回令牌")

print("\n【12】服务端模式（无本地存储，令牌由调用方保管）")
srv = LicenseClient(
    app_id="sdk_demo", server_url="http://127.0.0.1:3100",
    plugin_id=env["SDK_PLUGIN"], plugin_token=env["SDK_TOKEN"], plugin_secret=env["SDK_SECRET"],
    device_id="web-user-" + RUN, use_local_storage=False)
r = srv.activate(env["SDK_FRESH3"])
tok = r["license_token"]           # 这一步在真实项目里应该存进你的数据库
r2 = srv.verify(license_token=tok)
ok("服务端模式激活+验证通过，status=%s" % r2["status"])
try:
    LicenseClient(app_id="x", server_url="http://x", plugin_id="x", plugin_token="x",
                  plugin_secret="x", use_local_storage=False)
    bad("服务端模式缺 device_id 应当报错")
except ValueError as e:
    ok("缺 device_id 时明确报错: %s" % e)

print("\n\033[32m全部通过\033[0m  （共触发 %d 次策略回调）\n" % len(policies))

File: packages/sdk-php/tests/integration_test.php
<?php
declare(strict_types=1);
require __DIR__ . '/autoload.php';

use JcKami\ErrorCode;
use JcKami\LicenseClient;
use JcKami\LicenseError;

$env = [];
foreach (file('/tmp/sdk.env', FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES) as $line) {
    [$k, $v] = explode('=', $line, 2);
    $env[$k] = $v;
}

$policies = [];
$client = new LicenseClient([
    'app_id' => 'sdk_demo',
    'server_url' => 'http://127.0.0.1:3100',
    'plugin_id' => $env['SDK_PLUGIN'],
    'plugin_token' => $env['SDK_TOKEN'],
    'plugin_secret' => $env['SDK_SECRET'],
    'client_version' => '1.0.0',
    'on_policy' => function (array $p) use (&$policies) { $policies[] = $p; },
]);

function ok(string $m): void { echo "  \033[32m✓\033[0m $m\n"; }
function bad(string $m): void { echo "  \033[31m✗ $m\033[0m\n"; exit(1); }
function expectCode(callable $fn, string $code, string $label): void {
    try { $fn(); bad("$label 应当抛错但没有"); }
    catch (LicenseError $e) {
        $e->errorCode() === $code
            ? ok("$label: {$e->errorCode()} — {$e->getMessage()}")
            : bad("$label 期望 $code，实际 {$e->errorCode()}");
    }
}

echo "\n【1】设备指纹由稳定标识派生\n";
$run = bin2hex(random_bytes(4));
$device = LicenseClient::deviceIdFor('user:1001:' . $run);
ok('deviceIdFor("user:1001:...") = ' . substr($device, 0, 16) . '...');
if ($device !== LicenseClient::deviceIdFor('user:1001:' . $run)) bad('同一标识必须得到同一指纹');
ok('同一标识重复调用结果一致');
if ($device === LicenseClient::deviceIdFor('user:1002:' . $run)) bad('不同标识不应撞车');
ok('不同标识得到不同指纹');
try { LicenseClient::deviceIdFor('  '); bad('空标识应当抛错'); }
catch (\InvalidArgumentException $e) { ok('空标识被拒: ' . $e->getMessage()); }

echo "\n【2】拉取策略\n";
$p = $client->fetchPolicy($device);
ok("app_status={$p['app_status']}  kill_switch=" . var_export($p['kill_switch'], true));

echo "\n【3】激活\n";
$r = $client->activate($env['SDK_FRESH4'], $device, '网站用户 1001');
ok("license_id={$r['license_id']}  设备 {$r['device_count']}/{$r['device_limit']}  离线宽限 {$r['offline_grace_hours']}h");
$token = $r['license_token'];

echo "\n【4】验证（令牌滚动续期）\n";
$r2 = $client->verify($token, $device);
$r2['license_token'] !== $token ? ok('验证通过，换发了新令牌') : bad('令牌未变化');
$token = $r2['license_token'];

echo "\n【5】状态查询不返回令牌\n";
$s = $client->status(['license_key' => $env['SDK_FRESH4'], 'device_id' => $device]);
ok("status={$s['status']}  本机已绑定=" . var_export($s['device_bound'], true));
if (isset($s['license_token'])) bad('状态查询不应返回令牌');

echo "\n【6】错误卡密\n";
expectCode(fn() => $client->activate('ZZZZZ-ZZZZZ-ZZZZZ-ZZZZZ-ZZZZZ', $device),
    ErrorCode::LICENSE_NOT_FOUND, '不存在的卡密');

echo "\n【7】令牌与设备不匹配\n";
expectCode(fn() => $client->verify($token, LicenseClient::deviceIdFor('user:9999:' . $run)),
    ErrorCode::TOKEN_INVALID, '换个设备用同一令牌');

echo "\n【8】幂等键：重复提交只生效一次\n";
$d2 = LicenseClient::deviceIdFor('user:2002:' . $run);
$idem = 'php-idem-' . bin2hex(random_bytes(6));
$a = $client->activate($env['SDK_FRESH5'], $d2, null, $idem);
$b = $client->activate($env['SDK_FRESH5'], $d2, null, $idem);
$a['license_token'] === $b['license_token'] ? ok('第二次直接回放首次结果') : bad('幂等未生效');

echo "\n【9】解绑\n";
$d = $client->deactivate($b['license_token'], $d2, 'PHP SDK 测试解绑');
ok('已解绑，剩余换绑 ' . var_export($d['rebind_remaining'], true));
expectCode(fn() => $client->verify($b['license_token'], $d2),
    ErrorCode::TOKEN_INVALID, '解绑后旧令牌');

echo "\n【10】签名错误可被识别\n";
$badClient = new LicenseClient([
    'app_id' => 'sdk_demo', 'server_url' => 'http://127.0.0.1:3100',
    'plugin_id' => $env['SDK_PLUGIN'], 'plugin_token' => $env['SDK_TOKEN'],
    'plugin_secret' => 'sk_totally_wrong_secret', 'max_retries' => 0,
]);
expectCode(fn() => $badClient->fetchPolicy($device), ErrorCode::SIGNATURE_INVALID, '错误密钥');

echo "\n【11】网络不可达\n";
$offline = new LicenseClient([
    'app_id' => 'sdk_demo', 'server_url' => 'http://127.0.0.1:59999',
    'plugin_id' => $env['SDK_PLUGIN'], 'plugin_token' => $env['SDK_TOKEN'],
    'plugin_secret' => $env['SDK_SECRET'], 'max_retries' => 0, 'timeout_ms' => 1500,
]);
expectCode(fn() => $offline->fetchPolicy($device), ErrorCode::NETWORK_UNAVAILABLE, '连不上服务器');

echo "\n【12】缺少必填选项时明确报错\n";
try {
    new LicenseClient(['app_id' => 'x']);
    bad('应当抛 InvalidArgumentException');
} catch (\InvalidArgumentException $e) { ok('构造校验: ' . $e->getMessage()); }

echo "\n\033[32m全部通过\033[0m  （共触发 " . count($policies) . " 次策略回调）\n\n";
